159 uart dma update

Miceuz opened the pull request:
* Move serial event listeners from constructor of UART_DMA to particular startTX, startRX functions;  
* Expect uint8_t* buffers, not char* buffers as uint8_t is what we use 
* Rename to adhere to naming convention
* Remove stm32-test environment from platformio.ini as it does not fulfill its purpose after debug interface changes.



#439 Remove CRC32 implementation 
👉#440 uart dma update👈 **YOU ARE HERE**
#441 159 frame detector
#442 Add framing streams
#443 frame decoding
#444  Update comms to use UART_DMA and framing

hhff committed:
Implement gui build script and unit test harness

sglow committed:
Introduces the debug serial interface

sglow committed:
Working on driver for stepper motor controller

sglow committed:
Get/Set parameters working well

sglow committed:
Spinning a motor now.

sglow committed:
Minor edits testing multi-axis

sglow committed:
Ported to updated DMA registers in HAL

jlebar committed:
Templatize CircularBuffer's datatype.

GPC: circular-buffer-template

sglow committed:
Changed DMA channel used for steppers

jlebar committed:
C++ nits in uart_dma.

- "inline" should only be used on functions in header files.  In modern
  C and C++, it is a directive to the linker that a function should be
  deduplicated.  See https://en.cppreference.com/w/cpp/language/inline,
  "The original intent of the inline keyword..."

  Functions which are "private" to .cpp files should be marked static.
  This prevents an ODR violation (undefined behavior) which would occur
  if a different cpp file defined a function of the same name.

- Marked a one-arg constructor as "explicit".  Most one-arg constructors
  should be explicit; see https://en.cppreference.com/w/cpp/language/explicit.

- Got rid of `const uint32_t` in function declaration.  This is a nit,
  and a weird corner-case in C++.  Behold.

  The intent of this `const` is to prevent the definition of the
  function from modifying the const variable.  But

   (a) The definition may or may not specify const; it doesn't actually
       matter what the declaration says!

   (b) We could say, the declaration should match the definition.  But
       the caller cannot observe whether or not the callee modifies an
       argument that's passed by value.  Because it's not observable, it
       shouldn't be part of the API, i.e. it shouldn't be in the header.

GPC: dma-nits

sglow committed:
Added some functions to get/set motor parameters

jlebar committed:
Merge pull request #343 from RespiraWorks/jlebar/dma-nits

C++ nits in uart_dma.

jlebar committed:
Merge pull request #344 from RespiraWorks/jlebar/circular-buffer-template

Templatize CircularBuffer's datatype.

sglow committed:
Added more movement commands

sglow committed:
Disabled debug stuff unintentionally left on

jkff committed:
Merge pull request #308 from RespiraWorks/bug-report-template: Tweak bug report template

jkff committed:
Merge pull request #309 from RespiraWorks/feature-template: Tweak feature request template

sglow committed:
Support for commanding motors from both threads

sglow committed:
Added one time init and status.  Removed debug.

sglow committed:
Added some missing initialization to constructor

sglow committed:
Changed SPI control registers to bit-mapped style

sglow committed:
Minor tweaks based on review feedback

sglow committed:
Minor changes based on feedback from review

sglow committed:
Merge pull request #351 from RespiraWorks/issue_325_motor_driver

Issue 325 motor driver

sglow committed:
Revised A/D sampling method for improved readings

sglow committed:
Minor tweaks based on review feedback

sglow committed:
Minor changes to comments

sglow committed:
Merge pull request #353 from RespiraWorks/adc_improvements

Revised A/D sampling method for improved readings

Commit:
add task github issue template

Commit:
when to create Feature request or Task

Commit:
Merge pull request #354 from ysz/feature/task

add task github issue template

jkff committed:
Deletes controller-side alarm code - alarms will be handled in the GUI

jkff committed:
Merge pull request #357 from jkff/rm-alarms: Deletes controller-side alarm code - alarms will be handled in the GUI

asmodai27 committed:
Add functionnality to units:
+= and -= operator for ArithScalar
New ModuleScalar template for averaging/applying a multiplying factor (e.g calibration of flow?) right now only for flow and volume
Define Volume = VolumetricFlow*Duration = Duration*VolumetricFlow and VolumetricFlow = Volume/Duration

jkff committed:
Merge pull request #362 from jkff/reintro-debug: (Re-)Introduces the debug serial interface

jkff committed:
Makes ./test.sh pass with clang.

This includes one bug (= instead of ==) and something that is most
likely undefined behavior but somehow worked: all the debug cmds
were declared as static variables of type "FooCmd" in their respective
sources, but then as "extern DebugCmd" (different type) in debug.cpp.

git-pr-chain: test-sh

jkff committed:
Puts command classes in separate headers from their respective abstractions.

Names all command headers with _cmd and adds include guards.

git-pr-chain: debug-cmd-headers

jkff committed:
Merge pull request #363 from RespiraWorks/jkff/test-sh: Makes ./test.sh pass with clang.

jkff committed:
Merge pull request #364 from RespiraWorks/jkff/debug-cmd-headers: Puts command classes in separate headers from their respective abstractions

jkff committed:
Cosmetic symbol renames in vars.h

git-pr-chain: debug-vars

jkff committed:
Changes debug vars to be adapters over uint32

asmodai27 committed:
adress jlebar review comments:
- merge ModuleScalar in ArithScalar
- make ArithScalar*float commute
- make Duration*Flow and Volume/Duration work with the unit() fonctions to stay independent from implementation details
- get rid of some useless parenthesis

asmodai27 committed:
Address reviews: clarify comments and improve tests coverage of float operations

sglow committed:
Added a couple handy debug vars for tracing

sglow committed:
Restored old (wrong) diameter.  Will fix later.

Fixing this requires fixing the tests which were created based
on the old diameter.  I'll tackle that in a future PR

sglow committed:
Minor changes based on feedback from review

sglow committed:
Merge pull request #377 from RespiraWorks/add_debug_vars

Added a couple handy debug vars for tracing

jkff committed:
Adds vars_test.cpp

jkff committed:
Add test for var cmd

jkff committed:
Fix warning

jkff committed:
Merge pull request #365 from RespiraWorks/jkff/debug-vars: Changes debug vars to be adapters over uint32_t and adds basic tests

twslankard committed:
build an armv7 (aka armhf?) image instead of arm64, since the RPi can… (#379)

* build an armv7 (aka armhf?) image instead of arm64, since the RPi can't run arm64

* Issue #252: use non arch-specific Docker base image

twslankard committed:
Tag the latest master GUI docker build with `latest` in addition to SHA1 to make dev's lives easier. (#380)

NOTE: we'll still need to have a conversation about how updates are supposed to work, and whether we want to create release tags.

a-vinod committed:
Cache latest gui docker image & spin up on boot (#361)

* Run GUI container on boot

TODO code to be commented out once a respiraworks/gui docker image is on dockerhub.

* Spin up GUI Docker container on boot

TODO code will be commented out once a gui docker image is pushed to docker hub so the latest gui image is cached into the pi upon creation of the raspbian image.

Need edit Dockerfile so that the gui container runs the gui executable within it upon spinning up the container

The end result should be something like: rPi boots into Openbox -> GUI docker image spun up -> gui executable within container is run

* Cache & spin up gui docker image on boot

* Added tag to respiraworks/gui image pull & autostart

* Remove SHA1 tag

sglow committed:
Several minor debug improvements

sglow committed:
Added debug variables to allow cycling without GUI

jkff committed:
Separates the request and response buffers in DebugCmd

This will make it easier to instead pass a request and
response protobuf.

git-pr-chain: debug-req-resp

martukas committed:
populated with more READMEs and moved some content from wiki

jkff committed:
Puts tracing into a class.

git-pr-chain: trace-class

martukas committed:
updated main readme and enabled pre-commit

jlebar committed:
Pass -fno-threadsafe-statics.

This lets us use `static` variables inside of functions with types that
have nontrivial initializers.  Without this, you'll get a mysterious
error of the form

  sp/hard/libnosys.a(sbrk.o): In function `_sbrk':
  sbrk.c:(.text._sbrk+0x18): undefined reference to `end'

which indicates that something is trying to call malloc.

martukas committed:
dma uart test documentation

martukas committed:
more utils docs and controller docs more truthy

martukas committed:
some notes on hardware for testing utilities

martukas committed:
just a wee little bug

jlebar committed:
Merge pull request #384 from RespiraWorks/jlebar/threadsafe-initializers

Pass -fno-threadsafe-statics.

jkff committed:
Merge pull request #382 from RespiraWorks/jkff/debug-req-resp: Separates the request and response buffers in DebugCmd

sglow committed:
Fixed bug from earlier merge which broke debug tools

Also made some minor tweaks based on review feedback

sglow committed:
Added a comment

sglow committed:
Merge pull request #381 from RespiraWorks/minor_debug_improvements

Minor debug improvements

sglow committed:
Added SetAmpAll()

sglow committed:
Added module to control pinch valves

sglow committed:
Moved watchdog timer init to after startup code.

The startup code can take a while to run and runs before the
high priority loop starts running.  The high priority loop
is where the watchdog timer is reset, so this created a condition
where the watchdog timer could time out before we start servicing
it.

sglow committed:
Make sure ints are enabled before initing steppers

The stepper motor driver uses interrupts during it's initilization
it's therefore imporant that interrupts are enabled before the
stepper motors are initilized

sglow committed:
Fixed reversed direction

sglow committed:
Removed a script that isn't useful to others

jkff committed:
Merge pull request #272 from RespiraWorks/issue_270_remove_travis: CI Chore: Remove travis from repo, closes #270

jkff committed:
Removes references to Travis

jkff committed:
Merge pull request #385 from RespiraWorks/jkff/trace-class: Puts tracing into a class.

jkff committed:
Do not reuse buffer index between request and response.

Also:
* Adds functions for testing the debug serial port to HAL
* Adds a basic test of the Debug interface (just two commands for now)

git-pr-chain: buf-index

jkff committed:
Centralizes trace management completely in the Trace class and adds tests.

git-pr-chain: trace-tests

jkff committed:
Merge pull request #389 from RespiraWorks/jkff/circleci-yml: Removes references to Travis

martukas committed:
simple fan test

martukas committed:
"wider fan ramp range, temporary hack fix to HAL init bug; updates #390"

Miceuz committed:
Add a temporary band aid for #393 to unbreak the system by disabling compiler optimizations

Miceuz committed:
Declare cmd_ptr_ a volatile pointer to volatile data to avoid compiler optimizing out busy wait in StepMotor::SenCmd. Fixes #393

paulovap committed:
[GUI] Add script to run gui on eglfs 1920 x 1080 @ 60 mode

This script run_eglfs.sh is added alongside a kms configuration
file to force egl to run in a specific resolution/refresh_rate.

The script also enable some Qt debugging information to make it
easy to troubleshoot eventual issues with graphics

Also added a small fps counter element so we can keep track of
performance degratations. It will be available only in debug
builds once we start making this distinction.

jlebar committed:
Update jlebar/pre-commit-hooks repo.

Fixes Windows, and fixes cases where the user's tempdir is a different
filesystem than their home directory.

GPC: pre-commit-update

martukas committed:
Merge pull request #398 from RespiraWorks/jlebar/pre-commit-update

Update jlebar/pre-commit-hooks repo.

jlebar committed:
Add ProtoTraits traits class.

This lets you generically access e.g. MaxSize for T, where T is one of
our protos.

Meant to be used in https://github.com/RespiraWorks/VentilatorSoftware/pull/346.

GPC: proto-traits

martukas committed:
changes in response to PR

martukas committed:
Merge pull request #383 from RespiraWorks/martin_cleanup

Improving in-repo documentation

sglow committed:
Fixed some issues with powerSTEP chip

martukas committed:
"uncommented stepper init after rebasing; updates #390"

paulovap committed:
[GUI] Add Style singleton and extract left-side-panel controls into reusable Item

On main.qml we had few hard-coded controls for "RR", "PEEP",
"PIP" and so on. This change extracts those into a reusable
element called StepCounter which allows user to increase
or decrease a value, one step at the time.

Aside from that, a singleton element Style is created
to hold all UI related configuration, such as themes,
fonts, margin sizes and so on.

sglow committed:
Minor edits based on review

sglow committed:
Merge pull request #404 from RespiraWorks/issue_378_pinch_valve

Issue 378 pinch valve

paulovap committed:
Merge pull request #406 from RespiraWorks/ppinheiro/StepCounter

[GUI] Add Style singleton and extract left-side-panel controls into r…

paulovap committed:
Merge pull request #399 from RespiraWorks/ppinheiro/fix_eglfs_perf

[GUI] Add script to run gui on eglfs 1920 x 1080 @ 60 mode

jlebar committed:
Build with -fno-exceptions.

GPC: fno-exceptions

jlebar committed:
Clean up build flags.

- Make stm32_build_flags variable, shared between the two STM32 builds.
  This causes `-e stm32-test` to run with -fstrict-volatile.

- Filter src_test out of `native` build, so `pio run -e native` now
  doesn't try to build src_test.  (This isn't a standard command,
  usually we do `pio test -e native`.)

GPC: clean-build-flags

jlebar committed:
Build with -fno-rtti.

Per this guide to disabling exceptions in embedded C++:
http://elegantinvention.com/blog/information/smaller-binary-size-with-c-on-baremetal-g/

GPC: fno-rtti

jlebar committed:
Add an implementation of __cxa_pure_virtual.

This (together with patches earlier in the queue) allows us to use pure
virtual functions, yay.

Also use them in vars.h.

GPC: pure-virtual

paulovap committed:
[GUI] Add initial implementation on Selection Mode Popup

Mode selection popup will allow us to change the ventilator
operation mode. The design and copies are still
under development and will be refined later.

The organization of the QML files was also changed for easy understanding:

* control/ folder keeps all Pure UI controls, like buttons and widgets
* modes/ folder keeps the implementation of all modes. Later this could be
broken down if it gets too complex.

Add stub modes: PressureAssistMode and SupportPressureMode for
testing the mode selection

jlebar committed:
Merge pull request #402 from RespiraWorks/jlebar/proto-traits

Add ProtoTraits traits class.

jlebar committed:
Merge pull request #410 from RespiraWorks/jlebar/clean-build-flags

Clean up build flags.

jlebar committed:
Merge pull request #411 from RespiraWorks/jlebar/fno-exceptions

Build with -fno-exceptions.

jlebar committed:
Merge pull request #412 from RespiraWorks/jlebar/fno-rtti

Build with -fno-rtti.

jlebar committed:
Merge pull request #413 from RespiraWorks/jlebar/pure-virtual

Add an implementation of __cxa_pure_virtual.

martukas committed:
"resolved problem after rebase; updates #390"

sglow committed:
Fixed stepper commands that take no data.

sglow committed:
Added a table to linearize the pinch valve output

Also replaced the startup homing to an automatic homing state machine
that can run from the high priority loop rather then the background
task.  This has two advantages:
- It allows multiple pinch valves to be homed at the same time
- It allows the valves to be disabled when the ventilator is not in use.
  That saves power and is easier on the stepper motors and their drivers

sglow committed:
Removed obsolete call to home pinch valve on startup

martukas committed:
"fixed after rebase; updates #390"

martukas committed:
"solenoid test added; updates #390"

martukas committed:
"simple stepper motor test; updates #390"

martukas committed:
"test.sh builds integration tests; updates #390"

martukas committed:
"added pinch valve test; updates #390"

martukas committed:
"added some comments to test code; updates #390"

martukas committed:
Merge pull request #420 from RespiraWorks/issue_390

Basic vertical integration tests

sglow committed:
Made flow table length agnostic

sglow committed:
Merge pull request #419 from RespiraWorks/issue_418_pinch_valve_linearize

Issue 418 pinch valve linearize

jkff committed:
Make it possible to have several tests

jkff committed:
Merge pull request #391 from RespiraWorks/jkff/buf-index: Do not reuse buffer index between request and response.

jkff committed:
Merge pull request #422 from RespiraWorks/jkff/takeover-pr-95: [Takeover of #95] Adds Unit Test Harness for GUI, closes #87

jkff committed:
Merge pull request #392 from RespiraWorks/jkff/trace-tests: Centralizes trace management completely in the Trace class and adds tests.

paulovap committed:
[GUI Tests] Simplify GUI sub-project to run tests.

So, the current implementation of tests could be a bit
cumbersome, because needed to create a new qmake project
for each test class. This PR simplifies the gui folder structure
to accommodate tests sub-project in a way to keep all tests in a
single folder.

paulovap committed:
Merge pull request #416 from RespiraWorks/ppinheiro/mode_selection_popup

[GUI] Add initial implementation on Selection Mode Popup

jlebar committed:
First pass at TV zeroing.

Uses PID to push tidal volume to zero on each breath.

The PID could use some tuning, but as a first pass, this works OK for
me.

GPC: tv-zeroing

jlebar committed:
Set default params for NO_GUI_DEV_MODE.

Rather than starting the ventilator off, set some default values for the
NO_GUI_DEV_MODE params.

To do this, we had to modify the DebugT classes to let you specify a
default value on construction.

GPC: debug-defaults

jlebar committed:
Refactor controller_debug's `debug` command.

- Instead of a toggle, you specify 'debug on' or 'debug off'.
- Declare a global DbgPrint function instead of using a function
  pointer.

GPC: refactor-debug

jlebar committed:
Sort includes in controller_debug.py.

GPC: sort-includes

jlebar committed:
controller_debug: Print help for trace.

When you run plain `trace`, print an error and also print out `trace
help`.

To accomplish this, I reformatted the do_trace docstring so there isn't
a bunch of leading whitespace.  This matches other docstrings in here.
I didn't change the content of the docstring in this change.

GPC: print-trace-help

jlebar committed:
Fix typo: seperate -> separate.

GPC: separate

jlebar committed:
Add readline to list of imports.

This makes the cmd module use readline for handling inputs, so things
like ctrl+A work.

GPC: readline

jlebar committed:
Use argparse for `trace download` and `console`.

Using argparse is less code and gives more helpful errors, but more
importantly it will allow us to parse more complicated subcommands.

GPC: argparse

jlebar committed:
Get rid of most "global" statements.

"global" is only necessary when you assign to a global variable before
you read it.  We only have one variable like this; all of the other
"global" statements are unnecessary.

GPC: globals

jlebar committed:
Get rid of crc16 global variable.

It's unnecessary; we can just use a function.

GPC: crc16-global

jlebar committed:
Add documentation for `exec` command.

GPC: exec-documentation

jlebar committed:
Tighten up trace documentation a bit.

GPC: trace-docs

Miceuz committed:
Remove CRC32 implementation as it proved to be not that useful as it complicates code reuse between GUI and Cycle Controller. Also remove legacy 16 crc implementation

Miceuz committed:
Move serial event listeners from constructor of UART_DMA to particular startTX, startRX methods; some minor interface changes, some renaming

paulovap committed:
[GUI] Add ParameterButton element

This change has the following additions:

* Updates the HeaderButtons according to the current design document[0].
* Creates a Button element to change mode parameters on different modes.
* Adds new fonts according to the current design document[0].
* Replace parameter buttons in all modes.

0 - https://xd.adobe.com/view/5816839b-fd33-41f2-5a21-4b218fb9867f-2e27/screen/e0f99113-e872-4d96-84e4-c463f61812f1/01-mainscreen-Pressure-assist-01?fullscreen

paulovap committed:
[GUI] Add parameter display element

Adding a ParameterDisplay element that will be used
to show current value for all the relevant parameters.

Also update the design of the modes to follow more closely
the new design recommendation

jlebar committed:
Merge pull request #427 from RespiraWorks/jlebar/debug-defaults

Set default params for NO_GUI_DEV_MODE.

jlebar committed:
Merge pull request #428 from RespiraWorks/jlebar/refactor-debug

Refactor controller_debug's `debug` command.

jlebar committed:
Merge pull request #429 from RespiraWorks/jlebar/sort-includes

Sort includes in controller_debug.py.

jlebar committed:
Merge pull request #430 from RespiraWorks/jlebar/print-trace-help

controller_debug: Print help for trace.

jlebar committed:
Merge pull request #431 from RespiraWorks/jlebar/separate

Fix typo: seperate -> separate.

jlebar committed:
Merge pull request #432 from RespiraWorks/jlebar/readline

Add readline to list of imports.

jlebar committed:
Merge pull request #433 from RespiraWorks/jlebar/globals

Get rid of most "global" statements.

jlebar committed:
Merge pull request #434 from RespiraWorks/jlebar/crc16-global

Get rid of crc16 global variable.

jlebar committed:
Merge pull request #436 from RespiraWorks/jlebar/exec-documentation

Add documentation for `exec` command.

jlebar committed:
Merge pull request #437 from RespiraWorks/jlebar/trace-docs

Tighten up trace documentation a bit.

jlebar committed:
Merge pull request #438 from RespiraWorks/jlebar/argparse

Use argparse for `trace download` and `console`.

jlebar committed:
Fix up docstrings.

- Make some comments into docstrings.
- Complete incomplete docstring on TraceDownload.
- Format docstrings so the first line isn't empty (PEP 257)

GPC: docstrings

paulovap committed:
[GUI] Add stub for device status information.

Adds UI for device status (clock & battery).

Clock works out-of-the-box on computers with real time clocks,
but it was not tested in a raspberry pi.

Getting information from battery still not implemented.

paulovap committed:
Merge pull request #423 from RespiraWorks/ppinheiro/gui-tests-simplify

[GUI Tests] Simplify GUI sub-project to run tests.

Miceuz committed:
Remove unexisting include directories from platformio.ini, remove stm32_test environment as it's served its purpose

jlebar committed:
Add arguments to `trace start` command.

You can now do e.g. `trace start --period 10 volume flow_diff`.

GPC: trace-start-args

jlebar committed:
Auto-detect port device is connected to.

GPC: port-autodetect

jlebar committed:
Tweak implementation of TraceDownload.

No functional change, just more comments and a different implementation.

GPC: trace-download

jlebar committed:
Label axes and lines in graph.

- Label x axis as seconds (and scale according to trace period).
- Label individual lines according to what we're graphing, e.g.
  "Tidal volume, ml".

GPC: labels

jlebar committed:
Revert "Removed time checks from PID library."

This reverts commit be574eaddb12501e9bfff7f74b8f7210dbe554af.

The time checks were removed from PID because we were only calling PID
from the precisely-timed control loop.  We tried to update the PID every
(say) 10ms, and we were confident we *would* call every 10ms, plus or
minus a few microseconds.  But if we happened to call with a delta of
9.99ms, the PID code would skip our sample!  That's bad.

However, for flow correction, we need our PID library to be able to
handle the case when the time interval is not constant.  In this case,
we *do* want time checks.

So in this change we add time checks back.  The next commit makes time
checking optional, so the PID library can work for both cases.

GPC: pid-time

jlebar committed:
Use /usr/bin/env python3 as shebang.

This way we pull whatever is the user's default python3, instead of
forcing /usr/bin/python3.

GPC: shebang

jlebar committed:
Add -c flag.

Lets you run a single command, e.g.

  controller_debug.py -c "trace start volume" && sleep 10 &&
  controller_debug.py -c "trace graph"

Currently, there's no way to do multiple commands except by multiple
invocations of the script.

GPC: dash-c

jlebar committed:
Add debug var for tidal volume.

GPC: tv-debug-var

jlebar committed:
Add two sample 120s volume traces.

One wiggles all around, the other is a straight line.  Why, I cannot
say.

GPC: volume-traces

jlebar committed:
Merge pull request #445 from RespiraWorks/jlebar/docstrings

Fix up docstrings.

jlebar committed:
Merge pull request #446 from RespiraWorks/jlebar/trace-start-args

Add arguments to `trace start` command.

jlebar committed:
Merge pull request #447 from RespiraWorks/jlebar/port-autodetect

Auto-detect port device is connected to.

jlebar committed:
Merge pull request #448 from RespiraWorks/jlebar/trace-download

Tweak implementation of TraceDownload.

jlebar committed:
Merge pull request #449 from RespiraWorks/jlebar/labels

Label axes and lines in graph.

jlebar committed:
Merge pull request #451 from RespiraWorks/jlebar/shebang

Use /usr/bin/env python3 as shebang.

jlebar committed:
Merge pull request #452 from RespiraWorks/jlebar/dash-c

Add -c flag.

jlebar committed:
Merge pull request #453 from RespiraWorks/jlebar/tv-debug-var

Add debug var for tidal volume.

jlebar committed:
Merge pull request #455 from RespiraWorks/jlebar/volume-traces

Add two sample 120s volume traces.

jkff commented on common/libs/framing/serial_listeners.h:
```diff
@@ -0,0 +1,32 @@
+#ifndef __SERIAL_LISTENERS_H
+#define __SERIAL_LISTENERS_H
+
+enum RxError_t {
```
Please rename to simply RxError, make this an `enum class`, and remove the RX_ERROR prefix

jkff commented on common/libs/framing/serial_listeners.h:
```diff
@@ -0,0 +1,32 @@
+class RxListener {
+public:
+  // Called on RX complete
+  virtual void onRxComplete() = 0;
```
Rename to PascalCase naming style: OnRxComplete etc.

jkff commented on common/libs/framing/serial_listeners.h:
```diff
@@ -0,0 +1,32 @@
+// NOTE: all callbacks are called from interrupt context!
+class RxListener {
+public:
+  // Called on RX complete
```
Suggest renaming and rephrasing:
OnRxCompletedWithoutMarker - called if the requested Rx operation completed successfully without encountering the marker character
OnRxCompletedWithMarker - called if the requested Rx operation completed due to encountering the marker character
OnRxError - called if the requested Rx operation encountered an error

Miceuz replied:
UART_DMA does not maintain state if markers are received or not, those events are unrelated from the point of view of UART_DMA.

I would rather keep names are they are because they reflect exactly what happens in UARD_DMA dirver.

jkff replied:
Okay, could you add more detailed comments though? (e.g. that we receive either exactly N chars, or M<N chars until a marker, or a timeout because we didn't receive at least N chars within the timeout?) Or, in case the more detailed comments about the same concepts are located elsewhere, then point to that place.

jkff commented on common/libs/framing/serial_listeners.h:
```diff
@@ -0,0 +1,32 @@
+class TxListener {
+public:
+  // Called on TX complete
+  virtual void onTxComplete() = 0;
```
Also rename both to PascalCase

jkff commented on controller/lib/hal/uart_dma.cpp:
```diff
@@ -96,16 +99,19 @@ bool UART_DMA::isRxInProgress() {
   dma->channel[txCh].config.enable = 0; // Disable channel before config
   // data sink
-  dma->channel[txCh].pAddr = reinterpret_cast<REG>(&(uart->txDat));
+  dma->channel[txCh].pAddr = (&(uart->txDat));
```
This can be simply `&uart->txDat`

jkff commented on controller/lib/hal/uart_dma.cpp:
```diff
@@ -125,32 +131,39 @@ void UART_DMA::stopTX() {
 
   // data source
-  dma->channel[rxCh].pAddr = reinterpret_cast<REG>(&(uart->rxDat));
+  dma->channel[rxCh].pAddr = &(uart->rxDat);
```
ditto

jkff commented on controller/lib/hal/uart_dma.cpp:
```diff
@@ -212,50 +225,62 @@ void UART_DMA::UART_ISR() {
     // TODO define logic if stopRX() has to be here
-    rxListener.onRxError(e);
+    if (rxListener) {
+      rxListener->onRxError(e);
```
Should there be a `return;` here?

jkff commented on controller/lib/hal/uart_dma.cpp:
```diff
@@ -212,50 +225,62 @@ void UART_DMA::UART_ISR() {
-    rxListener.onCharacterMatch();
+    if (rxListener) {
+      rxListener->onCharacterMatch();
+    }
```
Please add some comments to the beginning of this method why it handles only the error and character match, but not RxComplete.

jkff commented on controller/lib/hal/uart_dma.h:
```diff
@@ -49,18 +22,20 @@ class UART_DMA {
   uint8_t rxCh;
-  UART_DMA_RxListener &rxListener;
-  UART_DMA_TxListener &txListener;
+  RxListener *rxListener = 0;
```
Rename to `rx_listener_` and `tx_listener_`

jkff commented on controller/lib/hal/uart_dma.h:
```diff
@@ -49,18 +22,20 @@ class UART_DMA {
-           char matchChar)
-      : uart(uart), dma(dma), txCh(txCh), rxCh(rxCh), rxListener(rxl),
-        txListener(txl), matchChar(matchChar) {}
+           uint8_t rxCh, char matchChar)
```
Rename: rx_ch, match_char (member variables accordingly rx_ch_, match_char_)

jkff commented on controller/lib/hal/uart_dma.h:
```diff
@@ -83,7 +59,8 @@ class UART_DMA {
   // was setup.
 
-  bool startRX(const char *buf, uint32_t length, uint32_t timeout);
+  [[nodiscard]] bool startRX(const uint8_t *buf, uint32_t length,
```
The comment seems wrong - it is reception of **at most** length chars, right?

Miceuz replied:
Actually it is **exactly** length chars  :)

jkff replied:
Please edit the comment then.

I'm still not completely following how this works though. So, the line is idle, and then:
* If it stays idle for timeout, then it's a timeout error (and we call rx->OnRxError - please mention that in the comment)
* If it becomes active, then we start receiving data
* If we receive exactly length bytes, then it's good (and then what happens? We call rxl->OnRxComplete, right? Please say so in the comment)

But this leaves a few questions open:
* After the line is not idle, how long do we wait for receiving length bytes? Is it also "timeout", and if so, does it count from the moment you called StartRx, or from the moment the line stopped being idle?
* What happens if we receive less than length bytes within that timeframe?
* What happens if we receive more than length bytes within that timeframe? If we just receive the first length - what happens to the rest: do they count towards the next rx, or are they simply discarded?

Miceuz replied:
The comment is updated, sorry, didn't mention that. 

All those events are independent. We order the DMA controller to receive x number of bytes and it will stay receiving those bytes, when DMA transfers the ordered number of bytes it will call OnRxComplete. Character match event is not related to DMA transfer, whenever UART sees the character it will issue the OnCharMatch, DMA transfer is not affected in any way. Similarly with errors - UART will issue an error interrupt calling OnRxError whenever there is an error on the RX line (framing, overflow, timeout) DMA transfer will hang there and wait for bytes, and UART will also keep providing either bytes or error interrupts if those occur (except for timeout - that is reset after a stop bit, i.e. after each successfully received char).

To answer your questions:
 * If line stays idle for timeout after last stop bit received, `OnRxError(TIMEOUT)` is called, that's it, DMA is still waiting for ordered number of bytes to come, driver does not change the reception configuration. This does not mean that transfer is cancelled.
 * Last stop bit is the start of timeout window. I.e. timeout counter starts after last received character. I think I understood what bugs you - timeout is not for the full length ordered for reception, but for the rx line. I.e. if line stays idle for timeout, interrupt is issued. Timeout error is not bound to DMA in any way, it's there to inform the caller that the line is idle. 

It's not clear how useful having the timeout interrupt really is, because we can timestamp frames in Comms decoded by FrameDecoder and notice that we didn't have a frame for too long. I suggest we land the FrameDecoder and if we see that timeout is not used in a reasonable way, I will remove it from UART_DMA.

jkff commented on platformio.ini:
```diff
@@ -144,7 +130,6 @@ check_flags =
   ; The actual checks are defined in .clang-tidy.
   clangtidy: --checks='-*' --extra-arg-before=-xc++ --extra-arg-before=-std=c++17
 check_patterns =
-  common/include
```
Why this change?

Miceuz replied:
We don't have this directory, having this here poisons vs code include resolution. It has to be removed some time, this time is as good as any

jkff reviewed (commented):
Thanks! Please edit the PR title and description to be more descriptive. I also added @sglow as a reviewer because I don't feel qualified to review the hardware aspect of this.

Miceuz committed:
Update singly byte processing comment; simplify code splitting uint32_t to single bytes

Miceuz committed:
Simplify RxError naming

Miceuz committed:
Return from ISR if an error was detected, don't analyze isCharacterMatch

Miceuz committed:
Rename to adhere to naming standard

sglow reviewed (approved):
I normally like to test hardware related changes myself before I approve them, but am out of town for a few days and don't want to hold you up.  I trust that you have thoroughly tested these changes.

jlebar committed:
Highlight the y=0 gridline.

Makes it easier to see where zero is.

GPC: black-gridline

jlebar committed:
Units: Add missing `const` modifiers.

Some member functions weren't marked as const.

Also remove some unnecessary `inline`s.

GPC: units-const

paulovap committed:
[GUI] Update mode selection popup to new design

paulovap committed:
[GUI] Update design of the graphs

Update the design of the graphs to follow new recommendations.

The current implementation of the ScopeView has a very high CPU usage,
making it unable to hold updates at 60Hz. (Actually I can at 90%
CPU usage).

The main reason is the fact we are using two Charts, one To draw a line,
and the other to draw the area. Hopefully, with enough time we can
customize the graphs to make it more efficient.

jkff commented on controller/lib/hal/uart_dma.cpp:
```diff
@@ -191,71 +204,90 @@ static bool isRxError() {
 }
 
+// ISR handler for the DMA peripheral responsible for reception.
+// Calls OnRxError and OnRxComplete functions of the tx_listener_
```
Did you mean rx_listener_?

jkff commented on controller/lib/hal/uart_dma.h:
```diff
@@ -1,66 +1,42 @@
+  DMA_Regs *const dma_;
+  uint8_t tx_ch_;
+  uint8_t rx_ch_;
+  RxListener *rx_listener_ = 0;
```
Please use nullptr instead of 0 for pointers

jkff reviewed (changes requested):
Thanks Albertas! A couple more comments, mainly for my understanding. Hopefully once I understand, I'll be able to just approve :)

tapemaster committed:
Add sample test data for inspiratory effort detection

jlebar committed:
gitignore .pytype/

GPC: gitignore-pytype

Miceuz committed:
More comments, fix OVERRUN erro name, use nullptr instead of 0

paulovap committed:
Merge pull request #450 from RespiraWorks/ppinheiro/new_design

Update GUI elements to new design recommendation

jlebar committed:
Merge pull request #459 from RespiraWorks/jlebar/gitignore-pytype

gitignore .pytype/

jlebar committed:
Merge pull request #460 from RespiraWorks/jlebar/black-gridline

Highlight the y=0 gridline.

jlebar committed:
Merge pull request #461 from RespiraWorks/jlebar/units-const

Units: Add missing `const` modifiers.

jlebar committed:
Granularity of Time and Duration 1ms -> 1us.

Motivation: The high-priority loop runs once every few ms, so just a
timer granularity of 1ms is very coarse compared to what's happening in
the system.

GPC: timer-us

jlebar committed:
Build with -wno-error=sign-compare.

This warning is about doing (int)x < (unsigned int)y.

AFAICT this is all noise for us, and in fact clang doesn't even enable
it with -Wall.

GPC: sign-compare

jlebar committed:
Remove sample-time checking from PID.

Now PID recomputes every time you call Compute().  Don't want that?
Then don't call compute!

This is made possible by #465, which changes the timer granularity from
1ms to 1us.  Without that change, the clock is too granular to be used
for the high-priority loop, which runs every 10ms -- due to jitter, we
might read 9ms or 10ms when the actual values are more like 9.999ms or
10.0001ms, resulting in an amplification of error.

GPC: PID-no-sample-period

sglow committed:
Implemented PID control of pressure

jlebar committed:
Remove unnecessary DbgPrint param from GetResp function.

GPC: dbg-print-param

sglow committed:
Minor changes based on reviewer feedback

sglow committed:
Merge pull request #469 from RespiraWorks/issue_421

Implemented PID control of pressure

tapemaster committed:
Merge pull request #457 from RespiraWorks/inspiratory-attempt-sample-data

Add sample test data for inspiratory effort detection

jlebar committed:
Merge pull request #470 from RespiraWorks/jlebar/sign-compare

Build with -wno-error=sign-compare.

jlebar committed:
Merge pull request #471 from RespiraWorks/jlebar/dbg-print-param

Remove unnecessary DbgPrint param from GetResp function.

jlebar committed:
Merge pull request #462 from RespiraWorks/jlebar/pid-time

Revert "Removed time checks from PID library."

jlebar committed:
Merge pull request #465 from RespiraWorks/jlebar/timer-us

Granularity of Time and Duration 1ms -> 1us.

jlebar committed:
Merge pull request #463 from RespiraWorks/jlebar/tv-zeroing

First pass at TV zeroing.

jlebar committed:
Merge pull request #467 from RespiraWorks/jlebar/PID-no-sample-period

Remove sample-time checking from PID.

paulovap committed:
Merge pull request #464 from RespiraWorks/ppinheiro/update_graphs

[GUI] Update design of mode selection popup and graphs

jlebar committed:
Fix units errors in sensors_test.

In multiple places in sensors_test, we use a tolerance in one unit (e.g.
kPa) to compare two values in a different unit (e.g. cmh2o).

Fix this by introducing EXPECT_{FLOW,PRESSURE,VOLUME}_NEAR macros.

GPC: flow-sensors-test

jlebar committed:
Update to new 3/4" Venturi specifications.

GPC: new-venturi

jlebar committed:
Disable sign-compare by using -Wno.

Rather than disabling it via -Wno-error.  This way it's totally silent.

GPC: sign-compare-take-2

jlebar committed:
Move setpoint_pressure field into ControllerState.

It doesn't belong in ActuatorState because the setpoint is not an
actuator!

GPC: move-setpoint-pressure

jlebar committed:
Merge pull request #476 from RespiraWorks/jlebar/sign-compare-take-2

Disable sign-compare by using -Wno.

jlebar committed:
Merge pull request #474 from RespiraWorks/jlebar/flow-sensors-test

Fix units errors in sensors_test.

jlebar committed:
Move flow zeroing into TVIntegrator.

GPC: flow-zeroing

jlebar committed:
Use nullptr instead of 0 for a void* param.

GPC: nullptr

jlebar committed:
Sleep for a few seconds at startup.

Due to a hardware issue, the blower always spins up when the board
restarts.  Wait for the blower to spin down, otherwise the sensors will
be miscalibrated.

GPC: sleep-at-startup

sglow committed:
Adjusted PID tuning to make system more stable.

Also reset PID state when ventilator is off

jlebar committed:
Merge pull request #481 from RespiraWorks/jlebar/move-setpoint-pressure

Move setpoint_pressure field into ControllerState.

jlebar committed:
Merge pull request #482 from RespiraWorks/jlebar/sleep-at-startup

Sleep for a few seconds at startup.

jlebar committed:
Merge pull request #480 from RespiraWorks/jlebar/nullptr

Use nullptr instead of 0 for a void* param.

jlebar committed:
Move blower_fsm implementations into cpp file.

Main motivation is to allow us to declare constants and functions that
are "private" to the file.

GPC: move-fsm-impls

jlebar committed:
Rise from PIP to PEEP in 100ms.

Instead of going directly from PIP to PEEP, transition over 100ms.

Citation: https://respiraworks.slack.com/archives/C011CJQV4Q7/p1591763842312500?thread_ts=1591759016.310200&cid=C011CJQV4Q7

GPC: pip-rise-time

jlebar committed:
Units: Allow x / y to return a raw ratio.

In a future PR, I want to divide two Durations.

Also mark more operators as const, and test that they are in fact const.

Please forgive the unironic use of pointer-to-member-function.

GPC: units-ratio

sglow committed:
Added automatic probing of stepper chips

This also reverses the direction of motion of the pinch valves
to conform to the standard wiring diagrams.

jlebar committed:
Merge pull request #475 from RespiraWorks/jlebar/new-venturi

Update to new 3/4" Venturi specifications.

jlebar committed:
Fix formatting in pid.cpp.

Somehow escaped clang-format.

GPC: pid-formatting

sglow committed:
Minor edits based on review feedback

sglow committed:
Merge pull request #487 from RespiraWorks/autoprobe_steppers

Added automatic probing of stepper chips

sglow committed:
Minor improvements to data tracing

Timestamps are added to saved trace data
controller is queried to find sample time rather then hard code it

sglow committed:
Added support for the controller audio alarm

sglow committed:
Comment changes

sglow committed:
Minor edits based on review feedback

sglow committed:
Merge pull request #490 from RespiraWorks/add_buzzer_support

Added support for the controller audio alarm

sglow committed:
Spelling error

jlebar committed:
Bump gcc version to 9.2.1.

This lets us use designated initializers with fields which have
nontrivial constructors (e.g. std::optional).

This is still not a bleeding-edge gcc, so I think we should be OK.

GPC: gcc9

jlebar committed:
Remove trailing whitespace in sample-data/README.md

GPC: trailing-whitespace

jlebar committed:
Eagerly home pinch valves.

Instead of homing them when you first start ventilation and then
hard-disabling them when you stop (requiring re-homing before we begin
again).

Also fix a TODO in controller.cpp.

Note that the stepper motors consume a nontrivial amount of power in
this mode.

GPC: controller-todo

sglow committed:
Merge pull request #489 from RespiraWorks/trace_improvements

Minor improvements to data tracing

jlebar committed:
Merge pull request #494 from RespiraWorks/jlebar/pid-formatting

Fix formatting in pid.cpp.

jlebar committed:
Merge pull request #495 from RespiraWorks/jlebar/trailing-whitespace

Remove trailing whitespace in sample-data/README.md

sglow committed:
Added test of PID reset

sglow committed:
Merge pull request #484 from RespiraWorks/adjust_tuning

Adjusted PID tuning to make system more stable.

jlebar committed:
Merge pull request #483 from RespiraWorks/jlebar/flow-zeroing

Move flow zeroing into TVIntegrator.

jlebar committed:
Merge pull request #485 from RespiraWorks/jlebar/units-ratio

Units: Allow x / y to return a raw ratio.

jlebar committed:
Merge pull request #486 from RespiraWorks/jlebar/move-fsm-impls

Move blower_fsm implementations into cpp file.

jlebar committed:
Merge pull request #488 from RespiraWorks/jlebar/pip-rise-time

Rise from PIP to PEEP in 100ms.

jlebar committed:
Merge pull request #496 from RespiraWorks/jlebar/gcc9

Bump gcc version to 9.2.1.

jlebar committed:
Merge pull request #497 from RespiraWorks/jlebar/controller-todo

Eagerly home pinch valves

Miceuz committed:
Remove timeout feature from uart_dma driver as we don't really use it

Miceuz committed:
Remove more timeout. Stop RX after error as DMA transmission is configured to stop

lhmzhou commented on controller/lib/core/actuators.h:
```diff
@@ -21,15 +21,33 @@ limitations under the License.
 };
 
-void actuators_init(void);
+// Causes passed state to be applied to the actuators
```
Why was `void actuators_init(void);` removed?

lhmzhou commented on gui/app/main.qml:
```diff
@@ -9,6 +9,12 @@ ApplicationWindow {
     id: window
     visible: true
     width: 1024
```
Is there a reason to keep `width: 1024` and `height: 600`?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/RespiraWorks/Ventilator?pullRequest=440) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/RespiraWorks/Ventilator?pullRequest=440) before we can accept your contribution.<br/>**1** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: Miceuz<br/>:x: sglow<br/>:x: jlebar<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/RespiraWorks/Ventilator?pullRequest=440) it.</sub>
